### PR TITLE
Analyzer with SpdxDocument improvements

### PR DESCRIPTION
**Pull request by fb33**
Generate a sourceArtifact with a file document location and don't set ConcludedLicense when it's not needed

### CODE REVIEW

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:208]: I believe a better documentation would be:

"Return the concluded license to be used in ORT's data model, which expects a not present value to be null instead of NONE or NOASSERTION."

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:216]: You could shorten the whole function to

    pkg.licenseConcluded.takeIf { SpdxConstants.isPresent(it) }?.toSpdx()

and just inline it into the caller.

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:222]: The addition of this is a different topic and should be done in a separate commit with a separate commit message explaining the details.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (fb33) [reply analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:222]: Ok, I'll be careful next time

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:222]: No worries, this can happen, and we're one of those project which are really strict about this, also see https://github.com/oss-review-toolkit/ort/blob/master/CONTRIBUTING.md#git-history.

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:209]: Nit: Missing dot at the end of the sentence.

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:208]: Please squash this commit into the previous one, instead of addressing review comments in new commits.

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:28]: Commit message: The title is too long, it should not exceed 75 chars. If you have more to say (which generally is good), then please put that into the commit message's body text.

Also the commit message should be prefix with the "scope" of the commit. In this case probably "SpdxDocumentFile: ...".

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:212]: If you want to keep this one-liner as a function instead of inlining it, I'd propose a name that does not use "build". The concluded license is not really built up / generated here, but just taken, with some condition. I'd just call this `getConcludeedLicense()`.

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:199]: This change again should be squashed into the (separate) commit that introduces the `buildSourceArtifact()` function.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by tsteenbe — CHANGES_REQUESTED** [review]
Grammar in commit message: "In case of a file uri in the downloadLocation, it is used to define the sourceArtifcat. So the Scanner can use it in its download phasis"

**Comment by sschuberth** [comment]
> 
> 
> Grammar in commit message: "In case of a file uri in the downloadLocation, it is used to define the sourceArtifcat. So the Scanner can use it in its download phasis"

Also: s/uri/URI/, s/sourceArtifcat/sourceArtifact/, s/phasis/phase/

**Comment by fb33** [comment]
> Grammar in commit message: "In case of a file uri in the downloadLocation, it is used to define the sourceArtifcat. So the Scanner can use it in its download phasis"

Sorry for my english. Is that sentence right for you ?
"In the case of a uri file in the downloadLocation, it is used to define the Artifcat source. Thus, the scanner can use it in its download phase."

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (tsteenbe) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:220]: Per SPDX spec 2.1.1, [Package Download Location](https://github.com/spdx/spdx-spec/blob/v2.1.1/chapters/3-package-information.md) can be a uniform resource locator but the docs don't specify accepted formats - I am not sure `file:/` can be used for this field. Need to discuss this with my fellow SPDX maintainers in our weekly tech call.

**Review by tsteenbe — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:219]: Again `build` probably is a misleading verb here, as you're not building the artifact itself. And again I'd simply propose `get` instead.

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:220]: Personally, I'd be fine with committing this now, and eventually fixing it up later.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (fviernau) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:211]: @sschuberth @tsteenbe @mnonnenmacher : Shall we file a ticket for making "NONE" a valid value for the package concluded license in ORT?

Note that for license finding curations we already do allow that.

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (fviernau) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:220]: I think if one slash is included in the comparision, the another one should be included for the same reason. E.g. use `file://`.

File: analyzer/src/funTest/assets/projects/synthetic/spdx-project-expected-output.yml
Comment (fviernau) [inline analyzer/src/funTest/assets/projects/synthetic/spdx-project-expected-output.yml:31]: commit-message: 
- "is map" -> "is mapped"
- Please wrap lines after 75? chars

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (fviernau) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:27]: Commit-message: Please wrap lines after 75? chars.

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (fviernau) [inline analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:215]: I wonder how the scanner is supposed to use / download a `file://` URL.
Is it absolute, relative, relative to what?
Do we want to handle `file://` URLs in ORT?

@sschuberth @tsteenbe any thoughts?

**Review by fviernau — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:211]: I'm a bit undecided. On the one hand it would be consistent to align on SPDX expressions for "everything about licenses". On the other hand, checking for (not) `null` is much more convenient than checking for `isPresent()` / `isNotPresent()`. My hunch is that we should still align on SPDX expression, but I'd like to also hear @mnonnenmacher opinion on that once he's back.

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:215]: Using `file://` URIs here was actually something that I suggested in the chat as a work-around / first step towards supporting them in the downloader. I guess this simply is something we *need* to add support for in the context of project-style SPDX file that reference package directories within the same repository.

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:211]: I've filed #3122 so we can have the discussion there later on.

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (fb33) [reply analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:215]: My need is to be able to verify open source policy rules on C/C++ projects with a proprietary package manager. 
The problem is that the dependencies referenced in this package manager don't have VCS. They are only binaries (dll ...) or directories containing different files: .h, lib, dll ...
I quickly made a mapping downloadLocation - sourceArtifact
I'm not very satisfied with the use of the sourceArtifact in file:/ but it's a fast workaround at the moment and I'm trying to move forward like that.
Maybe a notion of folderArtifact on the ORT model side and/or a notion of dependency type (archive, folder) on the Spdx file side is missing?

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:222]: @fb33  this is not yet solved...

**Review by fviernau — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt
Comment (fb33) [reply analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:222]: Hello, I follow the same idea as in Downloader:267
```
        // Some (Linux) file URIs do not start with "file://" but look like "file:/opt/android-sdk-linux".
        val sourceArchive = if (pkg.sourceArtifact.url.startsWith("file:/")) {
            File(URI(pkg.sourceArtifact.url))
```
But also, if I use file:// (ex: `file://C:/workspace/temp/legacy-test/libs/deps1/mydll.dll`) it crashes with 
```
Exception in thread "main" java.lang.IllegalArgumentException: URI has an authority component
        at java.base/java.io.File.<init>(File.java:426)
        at org.ossreviewtoolkit.downloader.Downloader.downloadSourceArtifact(Downloader.kt:269)
```
But I have no error if I use `file:/C:/workspace/temp/legacy-test/libs/deps1/mydll.dll`

Comment (fviernau) [reply analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:222]: Have you tried `file:///C:/` ?

Comment (fb33) [reply analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:222]: Yes, you're right. It works with ///
But the comment in Download :  "for some (linux) file URIs..." is still correct ?

Comment (fviernau) [reply analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:222]: I have no clue what comment you mean.

Comment (fb33) [reply analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:222]: In downloader/src/main/kotlin/Downloader.kt line 267
```
 // Some (Linux) file URIs do not start with "file://" but look like "file:/opt/android-sdk-linux".
        val sourceArchive = if (pkg.sourceArtifact.url.startsWith("file:/")) {
            File(URI(pkg.sourceArtifact.url))
```

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt:222]: I'm fine with merging this as-is to document the idea. File-based "downloads" require more work anyway to work completely.

@fviernau if you agree please dismiss your review / approve.

**Review by sschuberth — APPROVED** [review]

**Review by fviernau — APPROVED** [review]

**Pull request merged**
